files/xml.py

```diff
@@ -0,0 +1,470 @@
+    match = tree.xpath(xpath, namespaces=namespaces)
+    match_xpaths = []
+    for m in match:
+        match_xpaths.append(tree.getpath(m))
+    match_str = json.dumps(match_xpaths)
+    msg = "selector '%s' match: %s" % (xpath, match_str)
+    finish(module, tree, xpath, namespaces, changed=False, msg=msg)
+
+def count(module, tree, xpath, namespaces):
+    """ Return the count of nodes matching the xpath """
+    hits = tree.xpath("count(/%s)" % xpath, namespaces=namespaces)
+    finish(module, tree, xpath, namespaces, changed=False, msg=int(hits), hitcount=int(hits))
+
+def is_node(tree, xpath, namespaces):
+    """ Test if a given xpath matches anything and if that match is a node.
+
+    For now we just assume you're only searching for one specific thing."""
+    if xpath_matches(tree, xpath, namespaces):
+        # OK, it found something
+        match = tree.xpath(xpath, namespaces=namespaces)
+        if type(match[0]) == lxml.etree._Element:
+            return True
+
+    return False
+
+def is_attribute(tree, xpath, namespaces):
+    """ Test if a given xpath matches and that match is an attribute
+
+An xpath attribute search will only match one item"""
+    if xpath_matches(tree, xpath, namespaces):
+        match = tree.xpath(xpath, namespaces=namespaces)
+        if type(match[0]) == lxml.etree._ElementStringResult:
+            return True
+    return False
+
+def xpath_matches(tree, xpath, namespaces):
+    """ Test if a node exists """
+    if tree.xpath(xpath, namespaces=namespaces):
+        return True
+    else:
+        return False
+
+def delete_xpath_target(module, tree, xpath, namespaces):
+    """ Delete an attribute or element from a tree """
+    try:
+        for result in tree.xpath(xpath, namespaces=namespaces):
+            if not module.check_mode:
+                # Get the xpath for this result
+                if is_attribute(tree, xpath, namespaces):
+                    # Delete an attribute
+                    parent = result.getparent()
+                    # Pop this attribute match out of the parent
+                    # node's 'attrib' dict by using this match's
+                    # 'attrname' attribute for the key
+                    parent.attrib.pop(result.attrname)
+                elif is_node(tree, xpath, namespaces):
+                    # Delete an element
+                    result.getparent().remove(result)
+    except Exception, e:
+        abort(module, "Couldn't delete xpath target: %s (%s)" % (xpath, str(e)))
+    else:
+        finish(module, tree, xpath, namespaces, changed=True)
+
+def set_target_children(module, tree, xpath, namespaces, children):
+    matches = tree.xpath(xpath, namespaces=namespaces)
+
+    # Create a list of our new children
+    children = children_to_nodes(module=module, children=children)
+    children_as_string = [lxml.etree.tostring(c) for c in children]
+
+    changed = False
+
+    def replace_children_of(match):
```

Comment:
If there's no need to make this module a closure or similar, it's better to make these private toplevel functions instead of nested functions.